### PR TITLE
add upsert related configs

### PR DESCRIPTION
**Pull request by yupeng9**
## Description
Part of a series of PRs for https://github.com/apache/incubator-pinot/issues/4261
Check this [doc](https://docs.google.com/document/d/1qljEMndPMxbbKtjlVn9mn2toz7Qrk0TGQsHLfI--7h8/edit#heading=h.lsfmyoyyxtgt) out for the new design

 - Added the primary key definition in the schema
 - Changed the upsert config in table config. For the first phase, only support the full update mode.

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
configurations related to upsert

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 1]
```diff
@@ -66,6 +66,9 @@
   private TimeFieldSpec _timeFieldSpec;
   private final List<DateTimeFieldSpec> _dateTimeFieldSpecs = new ArrayList<>();
   private final List<ComplexFieldSpec> _complexFieldSpecs = new ArrayList<>();
+  // names of the columns that used as primary keys
+  // TODO: add validation checks like duplicate columns and use of time column
+  private List<String> _primaryKeyColumns;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:71]: This should not be part of the schema, but the upsert config

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 2]
```diff
@@ -27,44 +27,22 @@
...
-        "'primaryKeyColumns' must be configured with exact one column");
-    Preconditions.checkArgument(StringUtils.isNotEmpty(offsetColumn), "'offsetColumn' must be configured");
-    Preconditions.checkArgument(StringUtils.isNotEmpty(validFromColumn), "'validFromColumn' must be configured");
-    Preconditions.checkArgument(StringUtils.isNotEmpty(validUntilColumn), "'validUntilColumn' must be configured");
-    _primaryKeyColumns = primaryKeyColumns;
-    _offsetColumn = offsetColumn;
-    _validFromColumn = validFromColumn;
-    _validUntilColumn = validUntilColumn;
+  public enum Mode {
+    FULL, PARTIAL, NONE
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:33]: Recommend not putting `NONE` as a mode. Upsert is disabled when the `UpsertConfig` is `null` in the table config (more intuitive IMO)

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java [hunk 3]
```diff
@@ -265,4 +266,9 @@ public void setIngestionConfig(IngestionConfig ingestionConfig) {
   public void setTierConfigsList(List<TierConfig> tierConfigsList) {
     _tierConfigsList = tierConfigsList;
   }
+
+  @JsonIgnore
+  public UpsertConfig.Mode getUpsertMode() {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java:271]: Not sure if this method is needed

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 1 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:71]: Primary key is an inherent attribute of the schema. And its usage is beyond the upsert. For example, it's also used in the join feature which is happening in parallel.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 2 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:33]: In the implementation, it's more clearer (and easier) to have a value to represent that upsert is disabled. Also, when UpsertConfig is null, it implies the mode is `NONE`

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java (hunk 3 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java:271]: yes, in the upcoming PRs

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 4]
```diff
@@ -99,6 +102,14 @@ public void setSchemaName(String schemaName) {
     _schemaName = schemaName;
   }
 
+  public void setPrimaryKeyColumns(List<String> primaryKeyColumns) {
+    _primaryKeyColumns = primaryKeyColumns;
+  }
+
+  public List<String> getPrimaryKeyColumns() {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:109]: (nit) put getter before setter for consistency

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 5]
```diff
@@ -558,6 +576,11 @@ public SchemaBuilder addComplex(String name, DataType dataType) {
       return this;
     }
 
+    public SchemaBuilder addPrimaryKeyColumns(List<String> primaryKeyColumns) {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:579]:
```suggestion
    public SchemaBuilder setPrimaryKeyColumns(List<String> primaryKeyColumns) {
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments

**Comment by Jackie-Jiang** [comment]
After some consideration, I still think it is better to wrap the primary key columns inside the `UpsertConfig` for the following reasons:
- The schema for the offline segment is re-constructed from the segment metadata, where primary key columns might not be maintained
- For regular use cases, there is no primary key concept in Pinot. Putting it in `UpsertConfig` can keep all related configs at the same place
- For the upcoming JOIN feature, we should have another config for the in-memory lookup table, where we can put the primary key there. This 2 features cannot be applied to the same table, and I don't see the benefit of keeping this info inside the schema

Since this is the first feature using the primary key concept, I suggest putting all the related configs at the same place for easier management. If we find this as a common concept in the future, we can move it then.

**Comment by yupeng9** [comment]
I don't think so. Strictly speaking, primary key is a constraint defined on the schema, just like other constraints over the schema such as a column is `NOT NULL` or a column is `UNIQUE`. This is well defined in the database literature.

If the consideration is to limit its usage, then we can add a checker to ensure it's defined only when Upsert mode is not `NONE`. 
However, for cleaniness (and also the ease of future development and structuring), the constraints shall be defined as part of the schema.

> After some consideration, I still think it is better to wrap the primary key columns inside the `UpsertConfig` for the following reasons:
> 
> * The schema for the offline segment is re-constructed from the segment metadata, where primary key columns might not be maintained
> * For regular use cases, there is no primary key concept in Pinot. Putting it in `UpsertConfig` can keep all related configs at the same place
> * For the upcoming JOIN feature, we should have another config for the in-memory lookup table, where we can put the primary key there. This 2 features cannot be applied to the same table, and I don't see the benefit of keeping this info inside the schema
> 
> Since this is the first feature using the primary key concept, I suggest putting all the related configs at the same place for easier management. If we find this as a common concept in the future, we can move it then.

**Comment by yupeng9** [comment]
> After some consideration, I still think it is better to wrap the primary key columns inside the `UpsertConfig` for the following reasons:
> 
> * The schema for the offline segment is re-constructed from the segment metadata, where primary key columns might not be maintained
This is fine for now. As upsert applies to real-time segment only. In the future, I think we shall consider adding the primary ey concept to offline segment too
> * For regular use cases, there is no primary key concept in Pinot. Putting it in `UpsertConfig` can keep all related configs at the same place
> * For the upcoming JOIN feature, we should have another config for the in-memory lookup table, where we can put the primary key there. This 2 features cannot be applied to the same table, and I don't see the benefit of keeping this info inside the schema
For fact-dim join, the primary key actually needs to be defined on the regular table.
> 
> Since this is the first feature using the primary key concept, I suggest putting all the related configs at the same place for easier management. If we find this as a common concept in the future, we can move it then.

**Comment by yupeng9** [comment]
@Jackie-Jiang thanks for taking a look. comments addressed.

**Pull request merged**